### ISSUE TITLE
[Feature request] stop using Ajax to load labels

### ISSUE THREAD

**Issue by aledeg** [FreshRSS/FreshRSS#2270]
Yesterday, I was working on #2087. Everything went well until I realized that labels are loaded on demand instead of be always there. This means that when using shortcuts, I need to wait the end of the query.
I can't think of any good reason why we couldn't have labels as soon as we load the entry.

**Comment by Alkarex** [comment]
The main reason for the dynamic behaviour is performance, as the SQL request server side would be slightly heavy (especially on tiny hardware such as Raspberry Pi and with a big database), at least compared to the other requests we currently have.
An easy change would be for instance to pre-populate labels when the article is opened, instead of waiting till we open the drop-down menu

**Comment by Alkarex** [comment]
To clarify a bit: getting the full list of labels is cheap, and that might be enough for your needs (shortcut to add a label); we could provide that list in the JavaScript `context` variable. The only slightly expensive part is to have the list of used labels per article.

**Comment by aledeg** [comment]
I think that we could have a fast query even with joins. It really depends on how the query is written and how the database is indexed.

**Comment by Alkarex** [comment]
Yes we can have a fast query, but it is still an expensive one with at least two joins to add to our main query, so the overhead is signifiant. We already have the corresponding indexes (maybe there are possible improvements). An example of HTTP request, which returns the articles with their labels, can be found in our Google Reader API `/api/greader.php/reader/api/0/subscription/list`. There, after comparing the performances of various approaches, I went for two distinct SQL queries and a merge in PHP:

https://github.com/FreshRSS/FreshRSS/blob/e91ef155d42f9624ef0e5d553b4e5fad70ba1bb5/p/api/greader.php#L532-L535

Any performance optimisation welcome there :-)
I can happily test various suggestions on my small server / big database setup, which is nice for performance monitoring.

But back to the original need, don't you think it is enough to get the list of all labels for implementing the shortcut feature? You do not actually need to know what labels the article already has, do you?

**Comment by aledeg** [comment]
We could have only the labels loaded even if the selected one are still retrieved asynchronously. Doing so, there is no need to hop to the label table and only retrieve the linked ids.

**Comment by Alkarex** [comment]
Yes, that is quite cheap to do. This can either be done server side (with the drawback of generating a bunch of additional data to transfer) or client side (at load time or after some specific events). When do you need this data to be available more precisely for the shortcut feature?

**Comment by Alkarex** [comment]
Here is a quick example client side https://github.com/FreshRSS/FreshRSS/pull/2271 which immediately load the list of all labels when calling `loadDynamicTags()` while waiting for the actual label states

**Comment by aledeg** [comment]
Thank you! Can't you just retrieve the tag list server side, the same way the shortcuts are fetched?

**Comment by Alkarex** [comment]
The shortcuts are made available in the JavaScript variable `context.shortcuts`. Is this what you mean?

```javascript
//That would enable:
const labels = context.labels;
//Indead of the current:
const labels = document.querySelectorAll('#sidebar > .tags > .tree-folder-items .item-title').map(function (a) { return a.text; });
```

If yes, then it is indeed a simple change.

**Comment by aledeg** [comment]
Something like that or even like the share options. They are the same for all entries. So we can push them at the same time.

**Comment by Alkarex** [comment]
Yes, we can do the same than for the share options if needed. There are a couple of drawbacks though:
* More data to transfer
* The checkbox state cannot be set to the `indeterminate` state in pure HTML (only from JavaScript), and I think that state would be good to use while waiting for the reception of the real state (checked or not checked)

**Comment by math-GH** [comment]
Any news here? Are we talking about feed tags or user set labels?

**Comment by aledeg** [comment]
User set labels.
